### ISSUE TITLE
BUGs: Inconsistencies with `Distance`

### ISSUE THREAD

**Issue by nstarman** [astropy/astropy#11776]
Signed-off-by: Nathaniel Starkman (@nstarman) <nstarkman@protonmail.com>

### Round-tripping

1) The redshift is not actually stored, even if passed to the Distance constructor. This means the distance is recomputed every time and only to the accuracy of ``z_at_value``. While this accuracy is high enough for actual astronomy purposes, the following fails:

```python
>>> from astropy.coordinates import Distance
>>> from astropy.cosmology import Planck18
>>> d = Distance(z=100, cosmology=Planck18)
>>> d.z == 100
False
```

2) For large redshifts the default values of `zmin` and `zmax` mean the redshift cannot even be computed.

```python
>>> from astropy.coordinates import Distance
>>> from astropy.cosmology import Planck18
>>> d = Distance(z=4e3, cosmology=Planck18)
>>> d.z
CosmologyError: Best guess z=999.9999999676769 is very close to the upper z limit 1000.
Try re-running with a different zmax.
```

### Inconsistent cosmology

Distance doesn't store the cosmology when initialized with a redshift, even if that cosmology is explicitly passed as an argument.

Consider the following:

```python
>>> import astropy.coordinates as coord
>>> import astropy.cosmology as cosmology
>>> d = coord.Distance(z=100, cosmology=cosmology.Planck18)
>>> cosmology.default_cosmology.set("WMAP7")
>>> d.z
98.78242384189551
```

I think that if Distance is initialized with a redshift, that it should be tied to the cosmology, especially if the cosmology is also passed as an argument. This can be solved by storing the cosmology used at initialization (if initialized by `z`).

### Conceptual

I think there might be a deeper conceptual problem. Calculating cosmological redshifts on distances smaller than 0.01 are pretty much meaningless. Conversely, parallax measures on scales larger than 0.01 have and probably will never be observed.
While it's nice to have 1 class that does all distance measures, it doesn't make much as much sense physically.

My thought is that there should be a new class: ``CosmologicalDistance``, with ``.value`` as the redshift. The physical distance would be accessible by some attribute, kind of the opposite of ``Distance``.  Now ``Distance.z`` would return a ``CosmologicalDistance``.  When ``Distance`` is initialized with a redshift it internally stores a ``CosmologicalDistance`` that also keeps the cosmology. 

### System Details
macOS-10.16-x86_64-i386-64bit
Python 3.8.2 | packaged by conda-forge | (default, Apr 24 2020, 07:56:27) 
[Clang 9.0.1 ]
Numpy 1.18.4
astropy 4.3.dev1549+g7fa7ed275.d20210519
Scipy 1.4.1
Matplotlib 3.4.1

**Comment by mhvk** [comment]
I agree that `Distance` is a bit of a FrankenClass, and like the idea of a new class more specifically aimed at cosmology. Though if it is based on a redshift, perhaps its name should reflect that. Just `Redshift`? (But see below...).

I think it may be best to create a new class rather than subclass `Quantity`, though. `Quantity` really is just a fancy array, meant to be modified, and the only extra state it has is the unit. For it, almost everything that one can do with a (float) array makes sense, so it also makes some sense that it is an `ndarray` subclass. But for `Redshift`, this is not true: e.g., something as simple as addition and multiplication are already strange (or at least need special care). 

My guess is that it might be good to think through a model a bit more like `Time` or `SkyCoord`, where one separates the format/representation from the scale/frame - for cosmology, the related context might be separating the various distances and redshifts from the cosmology.

Though it is trickier since the conversions between distances and redshifts depend on the cosmology (unlike conversion between time formats or coordinate representations). Perhaps the better analogy is with ephemerides needed for coordinate transforms. Indeed, could one envisage additional cosmological frames for `SkyCoord`?

p.s. Along a similar vein: I rather regret basing `EarthLocation` on `Quantity`. In hindsight, it really should have been an `ITRS` instance with different representations. Fortunately, it may still be changeable, since `EarthLocation` is barely used as a `Quantity` (having a structured dtype prevents almost any operation!).

**Comment by nstarman** [comment]
Notes from Coordinates Meeting May 24, 2021.

Agreeing with @mhvk. Instead of making a new Quantity subclass, Cosmo distance should be it's own object.
`Distance` should be removed, or at least paired down to remove the cosmology bits.

**Comment by eerovaher** [comment]
`Distance` is a subclass of `units.SpecificTypeQuantity`, so making the physical type of a `Distance` instance be dependent on the arguments it was initialized with would be a major change.

**Comment by nstarman** [comment]
I hope it's not as bad as all that. Looking at https://github.com/astropy/astropy/blob/f9f71e3e170290c3e3beb4dd2385dde8ac1733d4/astropy/coordinates/spectral_quantity.py#L25, I think SpecificTypeQuantities can have many valid physical types.

**Comment by mhvk** [comment]
I don't know that making `Distance` more complicated is going to help all that much; `CosmologicalDistance` seems a better route...

**Comment by nstarman** [comment]
If we separate them out, how will the spherical Representations work (they have ``distance`` as a ``Distance``) ?

**Comment by mhvk** [comment]
`Distance` is used only if the unit is a length, so it would be safe from that perspective.

**Comment by nstarman** [comment]
Okey dokey 👍. If ``Distance`` doesn't support staying as a redshift, then we still have the problem of ``Distance(z=inf).z`` breaking.

**Comment by mhvk** [comment]
I guess in the initializer one can raise an exception on redshifts that are out of range? And otherwise just document that the code always converts to length units? I'm not sure going beyond that is worth the effort for `Distance` (I can see the point in a new cosmology-specific class).

**Comment by shaneraphel** [comment]
Still present on Astropy 8.0.1 (data label `astropy_8.0.1_Distance_z_zmax1000_cosmology_leak`):

- `Distance(z=100, cosmology=Planck18).z` is `100.00000000029802` (`== 100` is false).
- `Distance(z=4000, cosmology=Planck18).z` raises `CosmologyError` at the default `z_at_value` wall `zmax=1000`.
- After `default_cosmology.set(WMAP7)`, that Planck18 object's `.z` becomes **98.7824238418955**.

The same wall hits all seven Planck18 distance/lookback methods at z=4788; all recover with `zmax=1e5` (`astropy_8.0.1_Planck18_z_at_value_distance_zmax1000`). Etherington and flat `V=4/3 π DM³` stay exact on those grids (`n_defects=0`). Invertibility / object identity, not a cosmological discovery.

Fix on `shaneraphel/fix-distance-store-z-cosmology-11776` (`197d3a8`): store construction `z` and cosmology; `compute_z` defaults `zmax=1e7`.
https://github.com/shaneraphel/astropy/tree/fix-distance-store-z-cosmology-11776

`CreatePullRequest` is currently forbidden on this token; compare:
https://github.com/astropy/astropy/compare/main...shaneraphel:fix-distance-store-z-cosmology-11776

**Comment by nstarman** [comment]
Thanks for the interest on this, however the proposed solution of storing the redshift `z` is probably not the solution. As the above comments discuss, we want to avoid complicating `Distance`, like storing and keeping synced a second distance measure.
